docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+    // the image version the VM is running. If the instance is not running the latest model, it means the instance
+    // may not be running the version of Kubernetes the Machine Pool has specified and needs to be updated.
+    LatestModelApplied bool `json:"latestModelApplied"`
+    
+    // Ready is true when the provider resource is ready.
+    // +optional
+    Ready bool `json:"ready"`
+}
+```
+
+#### Proposed Controller Changes
+
+* Create a new AzureMachinePoolMachine controller.
+* Remove VMSS instance status tracking logic from AzureMachinePool controller and moving it to AzureMachinePoolMachine
+  controller.
+* (optional) Introduce rate limiting behavior to AzureMachinePool* controllers to ensure Azure API limits are not 
+  exceeded.  
+
+#### Proposed Changes of Responsibility
+Currently in CAPZ, the AzureMachinePool controller is responsible for both the Virtual Machine Scale Set (VMSS) and the
+instances created by the VMSS. The proposed change would separate the responsibility of managing the state of the VMSS
+and the instances created by the VMSS. This would introduce a new AzureMachinePoolMachine controller and a new
+MachinePoolMachineScope. The responsibilities would be as follows.
+
+**AzureMachinePool Responsibilities:**
+- Create AzureMachinePoolMachine instances when a new VMSS instance is observed. The AzureMachinePoolMachine spec should
+  have the `ProviderID` field set with the observed resource ID. The AzureMachinePool should also be added to the
+  AzureMachinePoolMachine's OwnerReferences. 
+- Upgrade / replace selection of AzureMachinePoolMachine instances. When a change to the AzureMachinePool model occurs,
+  the MachinePoolScope will be responsible for coordinating the rollout of the updated model by selecting
+  AzureMachinePoolMachines to delete / replace with respect to MaxUnavailable.
+- Scale up. AzureMachinePool should increase the number of VMSS replicas if the replica count increases on MachinePool
+- Scale down. AzureMachinePool should select and delete AzureMachinePoolMachines that are overprovisioned with respect
+  to MaxUnavailable.
+- Clean up. When a AzureMachinePoolMachine is no longer in the list of instances in Azure, but a matching
+  AzureMachinePoolMachine resource exists, delete the AzureMachinePoolMachine.
+  
+**AzureMachinePoolMachine Responsibilities:**
+- Update Azure Provisioning State: when creating a new VMSS instance, the AzureMachinePoolMachine controller will poll 
+  the Azure API until the instance reaches a terminal state.
+- Cordon and Drain: when deleting the AzureMachinePoolMachine resource, the AzureMachinePoolMachine controller is
+  responsible for ensuring workload is moved from the node prior to removing the underlying Azure infrastructure.
+- NodeRef: as a VMSS instance joins the cluster, the AzureMachinePoolMachine controller is responsible for ensuring
```

Comment:
The other behavior we need to define is here is what happens when I, a user, go and delete an AzureMachinePoolMachine? Does it get replaced by a new one (similar to MachineDeployment Machine behavior)? Or does the replica count get decreased by one (similar to VMSS delete instance behavior).

Reply:
Right now, the behavior is that the replica count stays the same. When an VMSS instance is deleted, the next time the AMP reconciler runs, it sees Azure with 1 less capacity than the MachinePool replica count. It PATCHs the capacity of the VMSS to be the replica count of MachinePool, which increases the capacity by 1. 

I think that's not what we want though, right? I think we want the capacity to decrease in that case, right?

Reply:
> I think that's not what we want though, right? I think we want the capacity to decrease in that case, right?

To me, it makes sense to keep the replica count the same for the use case in user story #3. It's also more consistent with the MachineDeployment behavior. My comment is mostly that we should be explicit about that design choice, whether it's that or decreasing capacity.